### PR TITLE
[google_maps_flutter] add tile overlays

### PR DESCRIPTION
**Pull request by cyanglaz**
This PR implements the tile overlay support for the android and iOS plugin. 

This is a fork of https://github.com/flutter/plugins/pull/2077 (CC @otopba )
This PR contains style fixes and integration tests that weren't in https://github.com/flutter/plugins/pull/2077

The platform interface code exists for build purpose, and should be removed when https://github.com/flutter/plugins/pull/3418 lands and is published.

Fixes https://github.com/flutter/flutter/issues/25942

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: packages/google_maps_flutter/google_maps_flutter/android/build.gradle
Comment (cyanglaz) [inline packages/google_maps_flutter/google_maps_flutter/android/build.gradle:44]: Note: this is to use lambda.

File: packages/google_maps_flutter/google_maps_flutter/pubspec.yaml
Comment (cyanglaz) [inline packages/google_maps_flutter/google_maps_flutter/pubspec.yaml:11]: TODO: update to version dep before landing

File: packages/google_maps_flutter/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java
Comment (bparrishMines) [inline packages/google_maps_flutter/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:82]: nit: 
```suggestion
          "fromBytes should have exactly one argument, interpretTileOverlayOptions the bytes. Got: "
```

File: packages/google_maps_flutter/google_maps_flutter/test/fake_maps_controllers.dart
Comment (bparrishMines) [inline packages/google_maps_flutter/google_maps_flutter/test/fake_maps_controllers.dart:373]: I believe collection literals for sets has landed. I think you can replace the line and others below:

```suggestion
      return <TileOverlayId>{};
```

**Review by bparrishMines — COMMENTED** [review]
Generally looks good with only a few comments

File: packages/google_maps_flutter/google_maps_flutter/ios/Classes/FLTGoogleMapTileOverlayController.m
Comment (cyanglaz) [reply packages/google_maps_flutter/google_maps_flutter/ios/Classes/FLTGoogleMapTileOverlayController.m:38]: hmm i thought it'd be nice the have a line separator between the interface and the properties. :) Can we leave the line here?

Comment (cyanglaz) [reply packages/google_maps_flutter/google_maps_flutter/ios/Classes/FLTGoogleMapTileOverlayController.m:38]: For example: https://github.com/flutter/plugins/blob/master/packages/google_maps_flutter/google_maps_flutter/ios/Classes/JsonConversions.m#L8

File: packages/google_maps_flutter/google_maps_flutter/ios/Classes/GoogleMapController.m
Comment (cyanglaz) [reply packages/google_maps_flutter/google_maps_flutter/ios/Classes/GoogleMapController.m:312]: This is basically a way of checking if `tileOverlaysToAdd` is nil. It should always be Array tho as the type was set in dart. Although we can change it to `if (tileOverlaysToAdd != nil)` I'd like to keep this as it is more consistent with other places in the plugin code. 

I do get your point tho, We can clean up this in the future?

File: packages/google_maps_flutter/google_maps_flutter/test/fake_maps_controllers.dart
Comment (cyanglaz) [reply packages/google_maps_flutter/google_maps_flutter/test/fake_maps_controllers.dart:373]: This min dart sdk in this plugin is 2.1 so we can't use it yet. Let's keep this as is in this plugin and have a separate PR to remove everything to use collection literal?

**Review by cyanglaz — COMMENTED** [review]
Thanks for the review. I've updated the code based on the comments. PTAL

File: packages/google_maps_flutter/google_maps_flutter/ios/Classes/FLTGoogleMapTileOverlayController.m
Comment (bparrishMines) [reply packages/google_maps_flutter/google_maps_flutter/ios/Classes/FLTGoogleMapTileOverlayController.m:38]: Yea, consistency is fine with me!

File: packages/google_maps_flutter/google_maps_flutter/ios/Classes/GoogleMapController.m
Comment (bparrishMines) [reply packages/google_maps_flutter/google_maps_flutter/ios/Classes/GoogleMapController.m:312]: If it's consistent with the rest of the plugin, then I'm fine with keeping the way it is in this PR

File: packages/google_maps_flutter/google_maps_flutter/test/fake_maps_controllers.dart
Comment (bparrishMines) [reply packages/google_maps_flutter/google_maps_flutter/test/fake_maps_controllers.dart:373]: It can probably be changed in the null safety migration or right after.

**Review by bparrishMines — APPROVED** [review]
LGTM pending tests

**Pull request merged**
